File: portals/api-control-plane/bff/internal/config/runtime_config.go
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package config

// buildRuntimeConfig collects the values the SPA reads from
// window.__RUNTIME_CONFIG__ (see src/config/runtime.ts's fromWindow()). Unlike
// ai-workspace's BFF — which invented its own APIP_AIW_* runtime-config
// vocabulary because its SPA was written alongside it — api-control-plane's
// SPA already has an established key vocabulary (camelCase, e.g.
// "platformApiBaseUrl") predating this BFF. This bridges to THOSE exact names
// rather than introducing a second convention the frontend would need to
// learn.
//
// Every backend call the SPA makes must go through this BFF's same-origin
// proxy — the browser never holds a token — so platformApiBaseUrl is always
// forced to the configured proxy prefix, never the upstream's real URL.
func buildRuntimeConfig(cfg *Config) map[string]string {
	out := map[string]string{
		"authMode":           cfg.Auth.Mode,
		"platformApiBaseUrl": cfg.ControlPlane.ProxyPrefix,
	}

	// billingProxyEnabled tells the SPA a "billing" named upstream exists, so
	// ProductActivation can call it (same-origin, via /proxy/billing/...)
	// without ever knowing the real billing service URL. Absent (defaults
	// false client-side) for every deployment that doesn't configure one —
	// every standalone deployment today.
	for _, u := range cfg.ControlPlane.Upstreams {
		switch u.Name {
		case "billing":
			out["billingProxyEnabled"] = "true"
		case "cloud":
			out["cloudProxyEnabled"] = "true"
		}
	}

	// Moesif wrap/basic iframe origin for cloud Insights embeds. Emitted
	// explicitly so the SPA never guesses from environmentName.
	if cfg.ControlPlane.MoesifAppURL != "" {
		out["moesifAppUrl"] = cfg.ControlPlane.MoesifAppURL
	}

	return out
}

File: portals/cloud-plugins/apip-cloud-ui-insights/src/config/runtimeConfig.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/// <reference path="./vite-env.d.ts" />

import { pickAllowlistedMoesifAppUrl } from '../utils/moesifEmbed';

export type InsightsRuntimeConfig = {
  /** Same-origin BFF proxy prefix, e.g. "/proxy". */
  platformApiBaseUrl: string;
  platformApiVersion: string;
  /** Moesif wrap host (iframe postMessage origin). Undefined when not configured. */
  moesifAppUrl?: string;
};

type LegacyWindowConfig = Partial<{
  PLATFORM_API_BASE_URL: string;
  platformApiBaseUrl: string;
  PLATFORM_API_VERSION: string;
  platformApiVersion: string;
  MOESIF_APP_URL: string;
  MOESIF_BASIC_INSIGHTS_URL: string;
  moesifAppUrl: string;
  moesifBasicInsightsUrl: string;
  /** AI Workspace BFF runtime key (`moesif_web_url` → APIP_AIW_MOESIF_WEB_URL). */
  APIP_AIW_MOESIF_WEB_URL: string;
}>;

const fromWindow = (): LegacyWindowConfig => {
  const globalWindow = window as Window & {
    config?: LegacyWindowConfig;
    __RUNTIME_CONFIG__?: LegacyWindowConfig;
  };
  return {
    ...(globalWindow.__RUNTIME_CONFIG__ ?? {}),
    ...(globalWindow.config ?? {}),
  };
};

const configuredMoesifAppUrl = () =>
  fromWindow().MOESIF_APP_URL ||
  fromWindow().MOESIF_BASIC_INSIGHTS_URL ||
  fromWindow().moesifAppUrl ||
  fromWindow().moesifBasicInsightsUrl ||
  fromWindow().APIP_AIW_MOESIF_WEB_URL ||
  import.meta.env.VITE_MOESIF_APP_URL ||
  import.meta.env.VITE_MOESIF_BASIC_INSIGHTS_URL ||
  import.meta.env.APIP_AIW_MOESIF_WEB_URL ||
  '';

const resolveInsightsMoesifAppUrl = (): string | undefined => {
  const configured = configuredMoesifAppUrl().trim();
  if (!configured) return undefined;
  return pickAllowlistedMoesifAppUrl(configured);
};

/** True when a trusted Moesif wrap origin is available (runtime or Vite env). */
export const isInsightsMoesifConfigured = (): boolean =>
  Boolean(resolveInsightsMoesifAppUrl());

/** Same-origin BFF proxy prefix from Vite `base` (`/` → `/proxy`, `/ai-workspace/` → `/ai-workspace/proxy`). */
export const defaultPlatformApiBaseUrl = (
  viteBase = import.meta.env.BASE_URL
): string => {
  const base = String(viteBase ?? '/').replace(/\/$/, '');
  return `${base}/proxy`;
};

export const insightsRuntimeConfig: InsightsRuntimeConfig = {
  platformApiBaseUrl:
    fromWindow().PLATFORM_API_BASE_URL ||
    fromWindow().platformApiBaseUrl ||
    import.meta.env.VITE_PLATFORM_API_BASE_URL ||
    defaultPlatformApiBaseUrl(),
  platformApiVersion:
    fromWindow().PLATFORM_API_VERSION ||
    fromWindow().platformApiVersion ||
    import.meta.env.VITE_PLATFORM_API_VERSION ||
    'v0.9',
  moesifAppUrl: resolveInsightsMoesifAppUrl(),
};

export const platformApiRoot = () => {
  const base = insightsRuntimeConfig.platformApiBaseUrl.replace(/\/$/, '');
  const version = insightsRuntimeConfig.platformApiVersion.replace(/^\//, '');
  return `${base}/api/${version}`;
};
